include/boost/numeric/odeint/algebra/detail/extract_value_type.hpp

```diff
@@ -28,24 +31,20 @@ namespace odeint {
 namespace detail {
 
 template< typename S , typename Enabler = void >
-struct extract_value_type {};
-
-// as long as value_types are defined we go down the value_type chain
-// e.g. returning S::value_type::value_type::value_type
-
-template< typename S >
-struct extract_value_type<S , typename boost::disable_if< has_value_type<S> >::type >
+struct extract_value_type
 {
-    // no value_type defined, return S
     typedef S type;
 };
 
+// as long as value_types are defined we go down the value_type chain
+// e.g. returning S::value_type::value_type::value_type
+
 template< typename S >
 struct extract_value_type< S , typename boost::enable_if< has_value_type<S> >::type >
-{
-    // go down the value_type
-    typedef typename extract_value_type< typename S::value_type >::type type;
-};
+  : mpl::if_< is_same< S, typename S::value_type > ,
+        mpl::identity< S > , // cut the recursion if S and S::value_type are the same
+        extract_value_type< typename S::value_type > >::type
```

Comment:
MPL is know to be slow to compile and work was already done to remove dependencies on it but here a new one is introduced, why not use C++11 typetraits (or the boost versions)?
E.g. something like:
`conditional<is_same<S, S::value_type>, common_type<S>, extract:value_type<S::value_type>>::type`

(omitted namespaces and `typename` for brevity)

Reply:
I pulled that straight out of a PR that had been sitting here a while. The immediate goal was to get the library correctly working again since Boost.Multiprecision support had been broken for years. Now that everything is working again I am going to spend time deliberately replacing all the C++03 components like MPL (which was already a dependency elsewhere).